spark/src/main/scala/org/apache/spark/sql/delta/GenerateIdentityValues.scala

```diff
@@ -0,0 +1,158 @@
+    ev.copy(code = code"""
+        final ${CodeGenerator.javaType(dataType)} ${ev.value} = $generatorTerm.next();
+        """, isNull = FalseLiteral)
+  }
+
+  /**
+   * Returns the [[DataType]] of the result of evaluating this expression.  It is
+   * invalid to query the dataType of an unresolved expression (i.e., when `resolved` == false).
+   */
+  override def dataType: DataType = LongType
+}
+
+object GenerateIdentityValues {
+  def apply(start: Long, step: Long, highWaterMarkOpt: Option[Long]): GenerateIdentityValues = {
+    new GenerateIdentityValues(PartitionIdentityValueGenerator(start, step, highWaterMarkOpt))
+  }
+}
+
+/**
+ * Generator of IDENTITY value for one partition.
+ *
+ * @param start The configured start value for the identity column.
+ * @param highWaterMarkOpt The optional high watermark for the identity value generation. If this is
+ *                      None, that means that no identity values has been generated in the past and
+ *                      we should start the identity value generation from the `start`.
+ * @param step IDENTITY value increment.
+ */
+case class PartitionIdentityValueGenerator(
+    start: Long,
+    step: Long,
+    highWaterMarkOpt: Option[Long]) {
+
+  require(step != 0)
+  // The value generation logic requires high water mark to follow the start and step configuration.
+  highWaterMarkOpt.foreach(highWaterMark => require((highWaterMark - start) % step == 0))
+
+  private lazy val base = highWaterMarkOpt.map(Math.addExact(_, step)).getOrElse(start)
+  private var partitionIndex: Int = -1
+  private var nextValue: Long = -1L
+  private var increment: Long = -1L
+
+
+  def initialize(partitionIndex: Int): Unit = {
+    if (this.partitionIndex < 0) {
+      this.partitionIndex = partitionIndex
+      this.nextValue = try {
+        Math.addExact(base, Math.multiplyExact(partitionIndex, step))
+      } catch {
+        case e: ArithmeticException =>
+          IdentityOverflowLogger.logOverflow()
+          throw e
+      }
+      // Each value is incremented by numPartitions * step from the previous value.
+      this.increment = try {
+        // Total number of partitions. In local execution case where TaskContext is not set, the
+        // task is executed as a single partition.
+        val numPartitions = Option(TaskContext.get()).map(_.numPartitions()).getOrElse(1)
+        Math.multiplyExact(numPartitions, step)
+      } catch {
+        case e: ArithmeticException =>
+          IdentityOverflowLogger.logOverflow()
+          throw e
+      }
+    } else if (this.partitionIndex != partitionIndex) {
+      throw SparkException.internalError("Same PartitionIdentityValueGenerator object " +
+        s"initialized with two different partitionIndex [oldValue: ${this.partitionIndex}, " +
+        s"newValue: $partitionIndex]")
+
+    }
+  }
+
+  private def assertInitialized(): Unit = {
+    if (partitionIndex == -1) {
```

Comment:
Could we put this on the same line as the method declaration? It'd reduce one level of indentation